### DISCUSSION TITLE
Reading in multi-page TIF and saving individual files is really slow

### DISCUSSION THREAD

**Discussion by StarFleetSecurity** [dlemstra/Magick.NET#1222] [Help]
I have a 25-page TIF file (about 7MB total size) that I need to split and save into individual files.

I had some old .NET Framework code doing this as below in the OldTiffToTiffs function. Creating the memory stream from the byte array and creating the "TiffBitmapDecoder" in the first few lines would take about 50 ms. Each new frame could be saved and added as a byte array to the "result" list in about 200 ms. So total processing time was about 5 seconds.

I am trying to recreate this functionality with the NewTiffToTiffs funtion using MagickImage in .NET 6 (since System.Windows.Media.Imaging is not available here). The function works just fine, but is prohibitively slow. Just creating the "MagickImageCollection" takes 5 seconds alone. Executing the "ToByteArray" on each image in the collection takes about 1200 ms. If I take out the dithering and compression lines, it is much faster (about 80 ms for each), but I must do the fax compression because we must have it as one bit per pixel and I must do the dithering or the image quality is terrible.

Is there a reason it takes so much longer to load the original TIF with MagickImage? And is there any way to speed up the "ToByteArray" when the dithering and compression are used? Having users wait 30 seconds instead of 5 is no good.

Thanks,

Andrew

        private static IEnumerable<byte[]> OldTiffToTiffs(byte[] tiff)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();

            // split tiff.new Fast WPF way
            var result = new List<byte[]>();
            using (MemoryStream msTifIn = new MemoryStream(tiff))
            {
                TiffBitmapDecoder tifIn = new TiffBitmapDecoder(msTifIn, BitmapCreateOptions.PreservePixelFormat, BitmapCacheOption.None);
                watch.Stop();
                Console.WriteLine($"TiffBitmapDecoder took {watch.ElapsedMilliseconds} ms");
                watch.Reset();

                foreach (BitmapFrame frame in tifIn.Frames)
                {
                    TiffBitmapEncoder encoder = new TiffBitmapEncoder();
                    encoder.Compression = TiffCompressOption.Ccitt4; 
                    watch.Start();
                    encoder.Frames.Add(frame);
                    using (MemoryStream msOut = new MemoryStream())
                    {
                        encoder.Save(msOut);
                        result.Add(msOut.ToArray());
                    }
                    watch.Stop();
                    Console.WriteLine($"Save/ToArray took {watch.ElapsedMilliseconds} ms");
                    watch.Reset();
                }
            }
            return result;
        }

        private static IEnumerable<byte[]> NewTiffToTiffs(byte[] tiff)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();

            var result = new List<byte[]>();
            using (MagickImageCollection imgs = new MagickImageCollection(tiff))
            {
                watch.Stop();
                Console.WriteLine($"MagickImageCollection took {watch.ElapsedMilliseconds} ms");
                watch.Reset();
                foreach (MagickImage img in imgs)
                {
                    watch.Start();
                    if (img.Depth > 1)
                    {
                        img.OrderedDither("o8x8");
                        img.Settings.Compression = CompressionMethod.Fax;
                    }
                    result.Add(img.ToByteArray(MagickFormat.Tif));
                    watch.Stop();
                    Console.WriteLine($"ToByteArray took {watch.ElapsedMilliseconds} ms");
                    watch.Reset();
                }
            }

            return result;
        }

**Comment by dlemstra** [comment]
Which version of Magick.NET are you using?

**Reply by StarFleetSecurity** [reply]
Sorry, the below was the runtime version. The actual version is 11.3.0.0.

I'm using version 2.0.50727 of Magick.NET.Core and Magick.NET-Q8-x64

**Reply by StarFleetSecurity** [reply]
I'm going to be on vacation for 2 weeks starting next week, so I was wondering if you might have any information regarding this before then.  Thanks.

**Comment by Sc-Dz** [comment]
You can try to solve it this way

public static void Test(string filename)
        {
            try
            {
                #region Image read Settings,Ignore the layer
                TiffReadDefines tiffRead = new TiffReadDefines
                {
                    IgnoreTags = new string[] { "ignore-layers" }
                };

                MagickReadSettings readSettings = new MagickReadSettings(tiffRead);
                #endregion

                int FrameCount = 0;

                #region GetPagesCount
                using (MagickImageCollection magicks = new MagickImageCollection())
                {
                    magicks.Ping(filename, readSettings);

                    FrameCount = magicks.Count;
                }
                #endregion

                using (MagickImage magickimage = new MagickImage())
                {
                    for (int i = 0; i < FrameCount; i++)
                    {
                        readSettings.FrameIndex = i;

                        magickimage.Read(filename, readSettings);

                        string new_path = Path.Combine(Path.GetDirectoryName(filename), Path.GetFileNameWithoutExtension(filename) + i.ToString().PadLeft(3, '0') + ".tif");

                        magickimage.Write(new_path, MagickFormat.Tif);
                    }
                }
            }
            catch (Exception ex)
            {

            }
        }

**Reply by dlemstra** [reply]
Thanks for your reply @Sc-Dz. I think you mean this instead for the `TiffReadDefines`:

```c#
TiffReadDefines tiffRead = new TiffReadDefines
{
    IgnoreLayers = true
};
```
